### PR TITLE
config for bootstrap proxy_list

### PR DESCRIPTION
**Pull request by neonphog**
### Summary

- adds a stub network config for fetching proxy_list from the bootstrap server (doesn't actually fetch yet)

### TODO:
- ~~[ ] CHANGELOG(s) updated with appropriate info~~
- ~~[ ] Just before pressing the merge button, ensure new entries to CHANGELOG(s) are still under the _UNRELEASED_ heading~~

### CODE REVIEW

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor.rs [hunk 1]
```diff
@@ -149,9 +148,20 @@ impl KitsuneP2pActor {
...
+                }
+                KitsuneP2pTx2ProxyConfig::Bootstrap {
+                    bootstrap_url,
+                    fallback_proxy_url,
+                } => {
+                    conf.client_of_remote_proxy = ProxyRemoteType::Bootstrap {
+                        bootstrap_url,
+                        fallback_proxy_url,
+                    };
+                }
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/spawn/actor.rs:164]: Perhaps this is where you want to query the server for a proxy url, and just keep this as an Optional url based on the result?

File: crates/kitsune_p2p/proxy/src/tx2.rs [hunk 2]
```diff
@@ -851,8 +854,11 @@ impl ProxyEp {
                             break;
                         }
 
-                        let timeout = tuning_params.implicit_timeout();
-                        let _ = hnd.get_connection(proxy_url.clone(), timeout).await;
+                        if let Some(proxy_url) = client_of_remote_proxy.get_proxy_url().await {
```
Comment (maackle) [inline crates/kitsune_p2p/proxy/src/tx2.rs:857]: Is it intended to query the bootstrap server here?

**Review by maackle — COMMENTED** [review]
Just some questions about when to query the bootstrap server, wondering if should be done ASAP, or more lazily as you have it.

Also wondering if kitsune is really the place for this, is the bootstrap server kitsune-specific, or more Holochain-specific? Would other consumers of kitsune realistically be using a bootstrap server? If so then this is fine. If not, perhaps this logic belongs in Holochain, keeping kitsune config simpler?

File: crates/kitsune_p2p/kitsune_p2p/src/spawn/actor.rs (hunk 1 above)
Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/spawn/actor.rs:164]: @maackle - I think we need this to be a runtime thing. As network conditions are changing, or we're explicitly removing / adding proxy servers to the proxy_list, I think we want running nodes that get disconnected to re-query and update to keep nodes running, right?

File: crates/kitsune_p2p/proxy/src/tx2.rs (hunk 2 above)
Comment (neonphog) [reply crates/kitsune_p2p/proxy/src/tx2.rs:857]: Correct, `get_proxy_url()` will do the bootstrap query.

Comment (neonphog) [reply crates/kitsune_p2p/proxy/src/tx2.rs:857]: As we get more mature, we can do fancy things within there like pinging all the proxies that were returned by the query and selecting the one with the lowest latency.

**Comment by neonphog** [comment]
@maackle 

> Just some questions about when to query the bootstrap server, wondering if should be done ASAP, or more lazily as you have it.
>
> Also wondering if kitsune is really the place for this, is the bootstrap server kitsune-specific, or more Holochain-specific? Would other consumers of kitsune realistically be using a bootstrap server? If so then this is fine. If not, perhaps this logic belongs in Holochain, keeping kitsune config simpler?

I think it needs to be done lazily as I say above. As such, I think kitsune needs to be a part of it. And the bootstrap protocol is already a dependency of kitsune, as kitsune is currently responsible for publishing the agent_infos to the bootstrap service.

We can re-design if there's a reason to, but that's how it's currently set up.

File: crates/kitsune_p2p/proxy/src/tx2.rs [hunk 3]
```diff
@@ -278,7 +321,7 @@ impl ProxyEpHnd {
         sub_ep_hnd: EpHnd,
         logic_hnd: LogicChanHandle<EpEvent>,
         backoff: Backoff,
-        client_of_remote_proxy: Option<ProxyUrl>,
+        cur_proxy_url: Arc<parking_lot::RwLock<Option<ProxyUrl>>>,
```
Comment (freesig) [inline crates/kitsune_p2p/proxy/src/tx2.rs:324]: Maybe we can use `Share` so we don't get any weird `.read()` deadlocks that seem to be common when using naked locks?

**Review by freesig — CHANGES_REQUESTED** [review]
Looks good except the naked lock

File: crates/kitsune_p2p/proxy/src/tx2.rs (hunk 3 above)
Comment (neonphog) [reply crates/kitsune_p2p/proxy/src/tx2.rs]: :+1: done

**Review by freesig — APPROVED** [review]

**Review by maackle — APPROVED** [review]
OK, thanks for explaining the rationale

**Pull request merged**